### PR TITLE
Improved benchmark runner

### PR DESCRIPTION
**Pull request by gflegar**
This PR adds the preconditioner benchmark to the benchmark runner script and generally improves the structure of the script.

Benchmark configuration is done entirely using environment variables and the make target is changed to simply `make benchmark`. It is explained in more detail in the README.

@tcojean I guess we would also need to modify the CI configurations a bit to accommodate these changes?

## TODO:

- [x] merge #151 first

### CODE REVIEW

**Comment by tcojean** [comment]
@gflegar Indeed we will have to change some things, but I believe it would be mostly localized in the scripts rather than the CI itself, actually.

**Comment by gflegar** [comment]
With the new benchmark runner, I am finally able to repay the debt from #137 and asses the performance of cooperative groups in more complicated kernels - the block-Jacobi preconditioner generation (compute-bound) and application (memory-bound). The system matrices used are block-diagonal matrices with 50000 blocks, and different block sizes on the diagonal (represented on the x-axis). The maximum block size of the preconditioner was set to match the block size of the matrix. The experiments were run on UJI's P100 GPU. I did not compile the block-Jacobi kernels for additional block sizes (because I forgot about it :smile:), but only the ones we compile for by default (I think 1, 13, 16, 32).

Here is the comparison for `simple_apply` kernel using the new cg-based implementation vs the old intrinsic-based one:

image

And here is the compute-bound `generate` kernel of preconditioner generation:

image

Note that the second plot includes not only inversion, but also diagonal block extraction (though it does _not_ include diagonal block detection) and storage into the preconditioner data structure. Unfortunately, I don't have an equivalent plot for the old MAGMA-sparse based implementation, as we never show performance results for the complete preconditioner generation on synthetic benchmarks.

Another interesting plot is the performance of diagonal block detection. The y-axis does not actually show "GFlop/s", as that kernel does not do any arithmetic, so I used the number of nonzeros of the matrix as the "flop" value in the calculation:

image

From this, it is obvious that block detection is the main bottleneck in preconditioner generation.
I'll produce a bar plot with a breakdown of the costs next.

**Comment by gflegar** [comment]
And the runtime breakdown plot, as promised:

image

The inversion step gets more expensive with larger blocks, but is still several orders of magnitude less expensive than the block detection.

**Comment by hartwiganzt** [comment]
I think we should definitely run on V100 too.

File: README.md [hunk 1]
```diff
@@ -173,6 +173,93 @@ run the following from the build folder:
...
+```
+
+Since `make` sets any variables passed to it as temporary environment variables,
+the following shorthand can also be used:
+
+```sh
+make benchmark VARIABLE="value" ...
+```
+
+A combinaion of the above approaches is also possible (e.g. it may be useful to
```
Comment (hartwiganzt) [inline README.md]: combination

File: README.md [hunk 2]
```diff
@@ -173,6 +173,93 @@ run the following from the build folder:
...
+                generated block-diagonal matrices.
+*   `DRY_RUN={true, false}` - If set to `true`, prepares the system for the
+    benchmark runs (downloads the collections, creates the result structure,
+    etc.) and outputs the list of commands that would normally be run, but does
+    not run the benchmarks themselves. Default is `false`.
+*   `EXECUTOR={reference,cuda,omp}` - The executor used for running the
+    benchmarks. Default is `cuda`.
+*   `SEGMENTS=<N>` - Splits the benchmark suite into `<N>` segments. This option
+    is usefull for running the benchmarks on an HPC system with a batch
+    scheduler, as it enables partitioning of the benchmark suite and runing it
```
Comment (hartwiganzt) [inline README.md]: running

**Comment by hartwiganzt** [comment]
LGTM

File: benchmark/run_all_benchmarks.sh [hunk 3]
```diff
@@ -70,42 +89,179 @@ run_spmv_benchmarks() {
 # is completed, the backups and the results are combined, and the newest file is
 # taken as the final result.
 run_solver_benchmarks() {
-    cp $1 "$1.imd" # make sure we're not loosing the original input
+    [ "${DRY_RUN}" == "true" ] && return
+    cp "$1" "$1.imd" # make sure we're not loosing the original input
     ./solver/solver --backup="$1.bkp" --double_buffer="$1.bkp2" \
-                    --executor=${EXECUTOR} --solvers="cg,bicgstab,cgs,fcg" \
+                    --executor="${EXECUTOR}" --solvers="cg,bicgstab,cgs,fcg" \
                     --max_iters=10000 --rel_res_goal=1e-6 \
```
Comment (pratikvn) [inline benchmark/run_all_benchmarks.sh]: I think these settings should also be mentioned in the README.

File: README.md (hunk 2 above)
Comment (pratikvn) [inline README.md]: /s/runing/running

**Comment by gflegar** [comment]
@hartwiganzt to build and run the benchmarks on the V100:

```sh
git checkout precond_benchmark_runner
mkdir relese; cd release
cmake .. -DCMAKE_BUILD_TYPE=Release -DBUILD_REFERENCE=ON -DBUILD_CUDA=ON -DBUILD_BENCHMARKS=ON -DBUILD_TESTS=ON -DDEVEL_TOOLS=OFF
make -j20
make test
make benchmark BENCHMARK=preconditioner EXECUTOR=cuda SYSTEM_NAME="V100-SXM2"
```

**Comment by gflegar** [comment]
Also, if you want to compile block-Jacobi kernels for all sizes, instead of just the selected few, download
[this patch](https://github.com/ginkgo-project/ginkgo/files/2519576/compile_all_kernels.patch.txt) to the build folder and run `git apply compile_all_kernels.patch.txt` just after the `git checkout` command.

File: benchmark/run_all_benchmarks.sh (hunk 3 above)
Comment (gflegar) [reply benchmark/run_all_benchmarks.sh]: done

**Comment by gflegar** [comment]
Just in case you are wondering, in addition to significantly increasing compile times, here is what the patch from the last commend does to performance:

image

image

**Comment by pratikvn** [comment]
LGTM!

**Pull request merged**